src/FragmentRegistry/FragmentRegistry.php

```diff
@@ -0,0 +1,166 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2017 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\FragmentRegistry;
+
+use Contao\ContentProxy;
+use Contao\CoreBundle\DependencyInjection\Compiler\FragmentRegistryPass;
+use Contao\CoreBundle\Framework\FrameworkAwareInterface;
+use Contao\CoreBundle\Framework\FrameworkAwareTrait;
+use Contao\ModuleProxy;
+use Contao\PageProxy;
+
+/**
+ * Fragment registry.
+ *
+ * @author Yanick Witschi <https://github.com/toflar>
+ */
+class FragmentRegistry implements FragmentRegistryInterface, FrameworkAwareInterface
+{
+    use FrameworkAwareTrait;
+
+    /**
+     * @var array
+     */
+    private $fragments = [];
+
+    /**
+     * @var array
+     */
+    private $fragmentOptions = [];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addFragment(string $identifier, $fragment, array $options): FragmentRegistryInterface
+    {
+        $this->ensureBasicOptions($options);
+
+        // Overrides existing fragments with same identifier
+        $this->fragments[$identifier] = $fragment;
+        $this->fragmentOptions[$identifier] = $options;
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFragment($identifier)
+    {
+        return $this->fragments[$identifier];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getOptions($identifier): array
+    {
+        return $this->fragmentOptions[$identifier];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFragments(callable $filter = null): array
+    {
+        $matches = [];
+
+        foreach ($this->fragments as $identifier => $fragment) {
+            if (null !== $filter && !$filter($identifier, $fragment)) {
+                continue;
+            }
+
+            $matches[$identifier] = $fragment;
+        }
+
+        return $matches;
+    }
+
+    /**
+     * Maps new fragments that were registered properly in the fragment
+     * registry to old $GLOBALS arrays for BC.
+     */
+    public function mapNewFragmentsToLegacyArrays()
```

Comment:
This should be in an event listener class (acting on the `initializeSystem` hook). Especially because it only handles the core types, but the `FragementRegistry` handles "everything".